### PR TITLE
Load all thumbnails asynchronously

### PR DESCRIPTION
**Pull request by chris-allan**
Switches all thumbnail loading to use asynchronous bulk strategies and implements remote call cancellation.

The thumbnail loading behaviour should now be identical in user experience to the existing thumbnail plugin.

OMERO 5.4.6 with openmicroscopy/openmicroscopy#5746 is essential to ensure that no background event craziness is impacting Parade and its thumbnail loading.

/cc @emilroz

### CODE REVIEW

File: src/js/App.js [hunk 1]
```diff
@@ -18,18 +18,21 @@
 
 import React, { Component } from 'react';
 import DataContainer from './dataLoader/DataContainer'
+import ThumbnailLoader from './dataLoader/Thumbnail'
```
Comment (will-moore) [inline src/js/App.js]: Would be more consistent to name the file ```ThumbnailLoader.js``` so it matches the class name. Easier to find etc.

File: src/js/dataLoader/Thumbnail.js [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+                qs.stringify(params, { indices: false })
+            )
+        })
+            .then(successCallback, failureCallback)
+            .then((response) => {
+                let nextImageIds = imageIds.slice(
+                    thumbnailsBatch, imageIds.length
+                );
+                if (nextImageIds.length > 0) {
+                    return this.loadThumbnails(
```
Comment (will-moore) [inline src/js/dataLoader/Thumbnail.js]: So you're loading the next batch of thumbnails on success of previous call, whereas in webclient we simply fire off all the batch calls at the same time https://github.com/openmicroscopy/openmicroscopy/blob/develop/components/tools/OmeroWeb/omeroweb/webclient/static/webclient/javascript/ome.webclient.actions.js#L452.
Is that difference intentional?

**Comment by snoopycrimecop** [comment]
Conflicting PR. Removed from build [parade-merge#59](https://ci.openmicroscopy.org/job/parade-merge/59/). See the [console output](https://ci.openmicroscopy.org/job/parade-merge/59/consoleText) for more details.
Possible conflicts:
  - PR #31 will-moore 'ImageIcon maxHeight'
    - src/js/dataView/dataset/ImageIcon.js

--conflicts

**Comment by will-moore** [comment]
Testing locally it seems to be working nicely.
If you are happy with #31 and want to merge it and fix the conflict, then we can test on web-dev-merge.
(or could exclude #31 if not).

**Comment by will-moore** [comment]
Thumbnails loading fine for single page app.
I noticed that in the Table layout that thumbnails are still loading individually, so they are re-loaded when you switch to that layout.
And when you switch back to thumbnails layout they are batch-loaded again.

I haven't looked at the code in detail yet, but I guess you may want the thumbnail data to be in a higher component to that it's available to all 3 layouts?

File: src/js/dataView/Layout.js [hunk 3]
```diff
@@ -58,6 +62,32 @@ class Layout extends React.Component {
         this.setState({showDatasets: show});
     }
 
+    loadThumbnails() {
+        const imageIds = this.props.filteredImages.map(v => v.id)
+            .filter(v => !this.state.thumbnails[v]);
+        console.log("imageIds", imageIds);
```
Comment (will-moore) [inline src/js/dataView/Layout.js]: Maybe don't need this now, unless it's still useful for debugging?

**Comment by will-moore** [comment]
This is working very nicely now.
No re-loading of thumbnails when switching layouts and asyn loading is much faster.

**Comment by jburel** [comment]
Everything looks good after rebuilding
We will need to indicate that parade 0.1.0 works best with version of OMERO 5.4.6 or newer

**Comment by chris-allan** [comment]
Agreed.

I'll try to remember to add it to the README.

**Pull request merged**